core/auth_provider_okta.go

```diff
@@ -0,0 +1,311 @@
+
+	return nil
+}
+
+func (p *OktaAuthProvider) WireUpTo(c *Core) {
+	p.core = c
+}
+
+func (p *OktaAuthProvider) ReferencedTenants() []string {
+	ll := make([]string, 0)
+	for _, m := range p.Mapping {
+		ll = append(ll, m.Tenant)
+	}
+	return ll
+}
+
+func (p *OktaAuthProvider) Initiate(r *route.Request) {
+
+	var state = randstr.Hex(16)
+
+	var redirectPath string
+
+	q := r.Req.URL.Query()
+	q.Add("client_id", p.ClientID)
+	q.Add("response_type", "code")
+	q.Add("response_mode", "query")
+	q.Add("scope", "openid profile email groups")
+	q.Add("redirect_uri", p.DeploymentURI + "/auth/okta/redir") //ex: https://shield.starkandwayne.com/auth/okta/redir
+	q.Add("state", state)
+	// q.Add("nonce", nonce)
+
+	redirectPath = p.OktaDomain + "/oauth2/" + p.AuthorizationServer + "/v1/authorize?" + q.Encode()
+	//ex: "https://dev-xyz.okta.com/oauth2/default/v1/authorize?" 
+
+	r.Redirect(302, redirectPath)
+}
+
+func (p *OktaAuthProvider) HandleRedirect(r *route.Request) *db.User {
+	
+	// The login is achieved through the authorization code flow, where the user 
+	// is redirected to the Okta-Hosted login page. After the user authenticates 
+	// they are redirected back to the application with an access code that is 
+	// then exchanged for an access_token.
+	
+	// log.Debugf("DEBUG::r *route.Request is [GET /auth/okta/redir]")
+
+	q := r.Req.URL.Query()
+	q.Add("grant_type", "authorization_code")
+	q.Set("code", r.Param("code", ""))
+	q.Add("redirect_uri", p.DeploymentURI + "/auth/okta/redir") //ex: https://shield.starkandwayne.com/auth/okta/redir
+	q.Add("response_type", "token")
+
+	url := p.OktaDomain + "/oauth2/" + p.AuthorizationServer + "/v1/token?" + q.Encode()
+
+	authHeader := base64.StdEncoding.EncodeToString([]byte(p.ClientID + ":" + p.ClientSecret))
+
+	req, _ := http.NewRequest("POST", url, bytes.NewReader([]byte("")))
+	h := req.Header
+	h.Add("Authorization", "Basic "+authHeader)
+	h.Add("Accept", "application/json")
+	h.Add("Content-Type", "application/x-www-form-urlencoded")
+	h.Add("Connection", "close")
+	h.Add("Content-Length", "0")
+
+	client := &http.Client{}
+	resp, _ := client.Do(req)
+	body, _ := ioutil.ReadAll(resp.Body)
+	defer resp.Body.Close()
+	var exchange Exchange
+	json.Unmarshal(body, &exchange)
+
+	if exchange.Error != "" {
+		p.Errorf("exchange resulted in error %s: %s", exchange.Error, exchange.ErrorDescription)
+		return nil
+	}
+
+	// Now we succussfully exhchanged the auth code for an openID token which is stored in our struct "exchange"
+	// Next we verify the token(s) and extract our okta claims from it (i.e. groups)
+	// Goal is to get {name, account, and orgs} from this token to do the mapping. 
+
+	id_token, verificationError := p.verifyToken(exchange.IdToken, "id")
+	if verificationError != nil {
+		fmt.Println(verificationError)
+	}
+	access_token, verificationError := p.verifyToken(exchange.AccessToken, "access")
+	if verificationError != nil {
+		fmt.Println(verificationError)
```

Comment:
Shouldn't we return here as verification failed?

Reply:
Done